### PR TITLE
Adding basic authentication to Admin client

### PR DESCRIPTION
**Pull request by wild-endeavor**
This adds the ability for the canonical golang gRPC Flyte Admin client to authenticate against an IDP.  This client is used by the execution plane (Flyte Propeller) when talking to the control plane (Admin) when it needs to send events primarily, among others.

The go-oidc library is used to discover the correct token endpoint necessary to retrieve a token.  Using the client id and secret provided from the configuration, a `TokenSource` is made and a `PerRPCCredentials` object is made from that which will keep request a token from the IDP if nil, keep track of expiration, and send the token along with gRPC requests as a metadata field when making calls.

* Ran goimports which made a lot of changes to everything.  Did not check in any changes under `/gen`
* make generate also changed a lot of things.

### CODE REVIEW

File: clients/go/admin/client.go [hunk 1]
```diff
@@ -48,19 +54,71 @@ func GetAdditionalAdminClientConfigOptions(cfg Config) []grpc.DialOption {
...
+	logger.Infof(ctx, "Constructing Admin client with token endpoint %s", provider.Endpoint().TokenURL)
+
+	return provider.Endpoint().TokenURL, nil
+}
+
+// This retrieves a DialOption that contains a source for generating JWTs for authentication with Flyte Admin
+func getAuthenticationDialOption(ctx context.Context, cfg Config) (grpc.DialOption, error) {
+	var tokenURL string
+	tokenURL, err := getTokenEndpointFromIssuer(ctx, cfg.IssuerURL)
+	if err != nil || tokenURL == "" {
```
Comment (katrogan) [inline clients/go/admin/client.go]: to get this straight, even if getTokenEndpointFromIssuer returns an err that's fine if we can get the token url from the config?

File: clients/go/admin/token_source.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
+package admin
+
+import "golang.org/x/oauth2"
+import "context"
+
+// This class is here because we cannot use the normal "github.com/grpc/grpc-go/credentials/oauth" package to satisfy
+// the credentials.PerRPCCredentials interface. This is because we want to be able to support a different 'header'
+// when passing the token in the gRPC call's metadata. The default is filled in in the constructor if none is supplied.
+type TokenSourcePerCallCredential struct {
```
Comment (katrogan) [inline clients/go/admin/token_source.go]: Is TokenSourcePerCallCredential the final name?

Comment (wild-endeavor) [reply clients/go/admin/token_source.go]: unless you tell me what to replace it with.

File: clients/go/admin/client.go (hunk 1 above)
Comment (wild-endeavor) [reply clients/go/admin/client.go]: yeah.  is that bad?

Comment (wild-endeavor) [reply clients/go/admin/client.go]: I added a comment, pushed.  Also renamed the function and the config option from `IssuerURL` to `AuthorizationServerURL` which I think is more accurate.

File: clients/go/admin/config.go [hunk 3]
```diff
@@ -28,10 +28,8 @@ type Config struct {
 
 	// There are two ways to get the token URL. If the issuer url is provided, the client will try to use RFC 8414 to
 	// try to get the token URL. Or you can just specify it directly.
-	// Note that the IssuerURL option doesn't follow redirects yet!!! So you'll need to specify the endpoint directly
```
Comment (katrogan) [inline clients/go/admin/config.go]: is this no longer true?

File: clients/go/admin/token_source.go (hunk 2 above)
Comment (katrogan) [reply clients/go/admin/token_source.go]: I'll defer to @EngHabu

File: clients/go/admin/client.go [hunk 4]
```diff
@@ -48,19 +54,73 @@ func GetAdditionalAdminClientConfigOptions(cfg Config) []grpc.DialOption {
...
+}
+
+// This retrieves a DialOption that contains a source for generating JWTs for authentication with Flyte Admin.
+// It will first attempt to retrieve the token endpoint by making a metadata call. If that fails, but the token endpoint
+// is set in the config, that will be used instead.
+func getAuthenticationDialOption(ctx context.Context, cfg Config) (grpc.DialOption, error) {
+	var tokenURL string
+	tokenURL, err := getTokenEndpointFromAuthServer(ctx, cfg.AuthorizationServerURL)
+	if err != nil || tokenURL == "" {
+		logger.Infof(ctx, "No token URL found from configuration Issuer, looking for token endpoint directly")
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply clients/go/admin/client.go]: okay

File: clients/go/admin/token_source.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply clients/go/admin/token_source.go]: will refactor

File: clients/go/admin/config.go [hunk 5]
```diff
@@ -18,18 +18,39 @@ type Config struct {
 	MaxBackoffDelay       config.Duration `json:"maxBackoffDelay" pflag:",Max delay for grpc backoff"`
 	PerRetryTimeout       config.Duration `json:"perRetryTimeout" pflag:",gRPC per retry timeout"`
 	MaxRetries            int             `json:"maxRetries" pflag:",Max number of gRPC retries"`
+
+	// Auth can only be used if also running with a secure connection. If UseInsecureConnection is set to true, none
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply clients/go/admin/config.go]: can i do this later?

[the remark this answers is not in the document]
Comment (wild-endeavor) [reply clients/go/admin/config.go]: can i not?

File: clients/go/admin/config.go [hunk 6]
```diff
@@ -18,18 +18,39 @@ type Config struct {
...
+
+	// There are two ways to get the token URL. If the issuer url is provided, the client will try to use RFC 8414 to
+	// try to get the token URL. Or you can just specify it directly.
+	AuthorizationServerURL string `json:"authorizationServerUrl" pflag:",This is the URL to your IDP's authorization server'"`
+	TokenURL               string `json:"tokenUrl" pflag:",Your IDPs token endpoint"`
+
+	// See the implementation of the 'grpcAuthorizationHeader' option in Flyte Admin for more information. But
+	// basically we want to be able to use a different string to pass the token from this client to the the Admin service
+	// because things might be running in a service mesh (like Envoy) that already uses the default 'authorization' header
+	GrpcAuthorizationHeader string `json:"grpcAuthorizationHeader" pflag:",Custom metadata header to pass JWT"`
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply clients/go/admin/config.go]: changed

**Pull request merged**